src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs

```diff
@@ -346,83 +342,78 @@ public Assembly LoadFromPath(string originalAnalyzerPath)
-        internal string GetRealAnalyzerLoadPath(string originalFullPath)
+        internal (string OriginalAssemblyPath, string RealAssemblyPath)[] GetPathMapSnapshot()
         {
             CheckIfDisposed();
 
             lock (_guard)
             {
-                if (!_analyzerAssemblyInfoMap.TryGetValue(originalFullPath, out var tuple))
-                {
-                    throw new InvalidOperationException($"Invalid original path: {originalFullPath}");
-                }
-
-                return tuple is { } value ? value.RealAssemblyPath : originalFullPath;
+                return _realToOriginalPathMap.Select(x => (x.Value, x.Key)).ToArray();
             }
         }
 
-        internal (string OriginalAssemblyPath, string RealAssemblyPath)[] GetPathMapSnapshot()
+#if NET
+        /// <summary>
+        /// Return an <see cref="IAnalyzerAssemblyLoader"/> which does not lock assemblies on disk that is
+        /// most appropriate for the current platform.
+        /// </summary>
+        /// <param name="windowsShadowPath">A shadow copy path will be created on Windows and this value 
+        /// will be the base directory where shadow copy assemblies are stored. </param>
+        internal static IAnalyzerAssemblyLoaderInternal CreateNonLockingLoader(
+            string windowsShadowPath,
+            ImmutableArray<IAnalyzerPathResolver> pathResolvers = default,
+            ImmutableArray<IAnalyzerAssemblyResolver> assemblyResolvers = default,
+            System.Runtime.Loader.AssemblyLoadContext? compilerLoadContext = null)
         {
-            CheckIfDisposed();
+            pathResolvers = pathResolvers.NullToEmpty();
+            assemblyResolvers = assemblyResolvers.NullToEmpty();
 
-            lock (_guard)
+            if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                return new AnalyzerAssemblyLoader(
+                    pathResolvers,
+                    [.. assemblyResolvers, StreamResolver.Instance],
+                    compilerLoadContext);
+            }
+
+            // The shadow copy analyzer should only be created on Windows. To create on Linux we cannot use 
+            // GetTempPath as it's not per-user. Generally there is no need as LoadFromStream achieves the same
+            // effect
+            if (!Path.IsPathRooted(windowsShadowPath))
             {
-                return _analyzerAssemblyInfoMap
-                    .Select(x => (x.Key, x.Value?.RealAssemblyPath ?? ""))
-                    .OrderBy(x => x.Key)
-                    .ToArray();
+                throw new ArgumentException("Must be a full path.", nameof(windowsShadowPath));
             }
+
+            return new AnalyzerAssemblyLoader(
+                [.. pathResolvers, new ProgramFilesAnalyzerPathResolver(), new ShadowCopyAnalyzerPathResolver(windowsShadowPath)],
+                [.. assemblyResolvers, DiskResolver.Instance],
+                compilerLoadContext);
```

Comment:
same comments about doc'ing why these are appropriate resolvers to pass along.

Open question: is ProgramFilesAnalyzerPathResolver stateful?  curious why a new isntance is passed along instead of singleton (like with DiskResolver).